### PR TITLE
Changed default local model to nomic

### PR DESCRIPTION
**Pull request by hagen-danswer**

### CODE REVIEW

File: backend/Dockerfile.model_server [hunk 1]
```diff
@@ -21,10 +21,12 @@ RUN apt-get remove -y --allow-remove-essential perl-base && \
 RUN python -c "from transformers import AutoModel, AutoTokenizer, TFDistilBertForSequenceClassification; \
 from huggingface_hub import snapshot_download; \
 AutoTokenizer.from_pretrained('danswer/intent-model'); \
-AutoTokenizer.from_pretrained('intfloat/e5-base-v2'); \
+AutoTokenizer.from_pretrained('nomic-ai/nomic-embed-text-v1'); \
+AutoTokenizer.from_pretrained('nomic-ai/nomic-bert-2048'); \
```
Comment (hagen-danswer) [inline backend/Dockerfile.model_server]: to make this work while airgapped, you need to .from_pretrained and snapshot_download not only nomic-ai/nomic-embed-text-v1, but also nomic-ai/nomic-bert-2048.

It was hard to find the exact reasoning for this, but I'm pretty sure it has something to do with nomic-embed-text-v1 being built on top of nomic-bert-2048 and that it needs to run .py scripts located only in the the nomic-bert-2048 repo [here](https://huggingface.co/nomic-ai/nomic-bert-2048/tree/main)

File: backend/danswer/configs/model_configs.py [hunk 2]
```diff
@@ -34,8 +34,8 @@
 SIM_SCORE_RANGE_LOW = float(os.environ.get("SIM_SCORE_RANGE_LOW") or 0.0)
 SIM_SCORE_RANGE_HIGH = float(os.environ.get("SIM_SCORE_RANGE_HIGH") or 1.0)
 # Certain models like e5, BGE, etc use a prefix for asymmetric retrievals (query generally shorter than docs)
-ASYM_QUERY_PREFIX = os.environ.get("ASYM_QUERY_PREFIX", "query: ")
-ASYM_PASSAGE_PREFIX = os.environ.get("ASYM_PASSAGE_PREFIX", "passage: ")
+ASYM_QUERY_PREFIX = os.environ.get("ASYM_QUERY_PREFIX", "search_query: ")
+ASYM_PASSAGE_PREFIX = os.environ.get("ASYM_PASSAGE_PREFIX", "search_document: ")
 # Purely an optimization, memory limitation consideration
```
Comment (hagen-danswer) [inline backend/danswer/configs/model_configs.py:39]: these are the defaults for nomic-ai/nomic-embed-text-v1

File: backend/model_server/encoders.py [hunk 3]
```diff
@@ -190,7 +190,9 @@ def get_embedding_model(
 
     if model_name not in _GLOBAL_MODELS_DICT:
         logger.info(f"Loading {model_name}")
-        model = SentenceTransformer(model_name)
+        model = SentenceTransformer(
+            model_name_or_path=model_name, trust_remote_code=True
+        )
```
Comment (hagen-danswer) [inline backend/model_server/encoders.py:209]: This is related to the needing to also install nomic-bert-2048
there is a script that has to be executed to use the model (unsure when) that is located in nomic-bert-2048 and not in nomic-embed-text-v1 (a couple .py scripts you can see [here](https://huggingface.co/nomic-ai/nomic-bert-2048/tree/main))

Not 100% sure though

File: backend/danswer/natural_language_processing/utils.py [hunk 4]
```diff
@@ -116,8 +116,9 @@ def get_tokenizer(model_name: str | None, provider_type: str | None) -> BaseToke
         if provider_type.lower() == "openai":
             # Used across ada and text-embedding-3 models
             return _check_tokenizer_cache("openai")
+        # If we are given a cloud provider_type that isn't OpenAI, we default to trying to use the model_name
+        # this means we are approximating the token count which may leave some performance on the table
```
Comment (hagen-danswer) [inline backend/danswer/natural_language_processing/utils.py:121]: general note

File: backend/Dockerfile.model_server [hunk 5]
```diff
@@ -18,14 +18,20 @@ RUN apt-get remove -y --allow-remove-essential perl-base && \
...
 # Pre-downloading models for setups with limited egress
-RUN python -c "from transformers import AutoModel, AutoTokenizer, TFDistilBertForSequenceClassification; \
-from huggingface_hub import snapshot_download; \
-AutoTokenizer.from_pretrained('danswer/intent-model'); \
-AutoTokenizer.from_pretrained('intfloat/e5-base-v2'); \
-AutoTokenizer.from_pretrained('mixedbread-ai/mxbai-rerank-xsmall-v1'); \
-snapshot_download('danswer/intent-model'); \
-snapshot_download('intfloat/e5-base-v2'); \
-snapshot_download('mixedbread-ai/mxbai-rerank-xsmall-v1')"
+RUN python -c "from transformers import AutoTokenizer; \
```
Comment (yuhongsun96) [inline backend/Dockerfile.model_server:21]: It's better to combine these into a single layer. If you do a single RUN it creates a single layer that can be cached.

File: backend/model_server/encoders.py [hunk 6]
```diff
@@ -199,7 +199,10 @@ def get_embedding_model(
 
     if model_name not in _GLOBAL_MODELS_DICT:
         logger.info(f"Loading {model_name}")
-        model = SentenceTransformer(model_name)
+        model = SentenceTransformer(
+            model_name_or_path=model_name,
+            trust_remote_code=True,
```
Comment (yuhongsun96) [inline backend/model_server/encoders.py:208]: I would add a comment here:
"Some model architectures that aren't built into the Transformers or Sentence Transformer need to be downloaded to be loaded locally. This does not mean data is sent to remote servers for inference, however the remote code can be fairly arbitrary so only use trusted models"

**Pull request merged**